src/cpp/server/orca/backend_metric_recorder.cc

```diff
@@ -0,0 +1,291 @@
+  if (!IsQpsValid(value)) {
+    if (GRPC_TRACE_FLAG_ENABLED(grpc_backend_metric_trace)) {
+      gpr_log(GPR_INFO, "[%p] QPS rejected: %f", this, value);
+    }
+    return;
+  }
+  if (GRPC_TRACE_FLAG_ENABLED(grpc_backend_metric_trace)) {
+    gpr_log(GPR_INFO, "[%p] QPS set: %f", this, value);
+  }
+  update_seq_.fetch_add(1, std::memory_order_acq_rel);
+  qps_.store(value, std::memory_order_relaxed);
+}
+
+void ServerMetricRecorder::SetNamedUtilization(std::string name, double value) {
+  if (!IsUtilizationValid(value)) {
+    if (GRPC_TRACE_FLAG_ENABLED(grpc_backend_metric_trace)) {
+      gpr_log(GPR_INFO, "[%p] Named utilization rejected: %f name: %s", this,
+              value, name.c_str());
+    }
+    return;
+  }
+  internal::MutexLock lock(&mu_);
+  if (GRPC_TRACE_FLAG_ENABLED(grpc_backend_metric_trace)) {
+    gpr_log(GPR_INFO, "[%p] Named utilization set: %f name: %s", this, value,
+            name.c_str());
+  }
+  update_seq_.fetch_add(1, std::memory_order_relaxed);
+  named_utilization_[std::move(name)] = value;
+}
+
+void ServerMetricRecorder::SetAllNamedUtilization(
+    std::map<std::string, double> named_utilization) {
+  internal::MutexLock lock(&mu_);
+  update_seq_.fetch_add(1, std::memory_order_relaxed);
+  named_utilization_ = std::move(named_utilization);
+  if (GRPC_TRACE_FLAG_ENABLED(grpc_backend_metric_trace)) {
+    gpr_log(GPR_INFO, "[%p] All named utilization updated.", this);
+  }
+}
+
+void ServerMetricRecorder::ClearCpuUtilization() {
+  update_seq_.fetch_add(1, std::memory_order_acq_rel);
+  cpu_utilization_.store(-1.0, std::memory_order_relaxed);
+  if (GRPC_TRACE_FLAG_ENABLED(grpc_backend_metric_trace)) {
+    gpr_log(GPR_INFO, "[%p] CPU utilization cleared.", this);
+  }
+}
+
+void ServerMetricRecorder::ClearMemoryUtilization() {
+  update_seq_.fetch_add(1, std::memory_order_acq_rel);
+  mem_utilization_.store(-1.0, std::memory_order_relaxed);
+  if (GRPC_TRACE_FLAG_ENABLED(grpc_backend_metric_trace)) {
+    gpr_log(GPR_INFO, "[%p] Mem utilization cleared.", this);
+  }
+}
+
+void ServerMetricRecorder::ClearQps() {
+  update_seq_.fetch_add(1, std::memory_order_acq_rel);
+  qps_.store(-1.0, std::memory_order_relaxed);
+  if (GRPC_TRACE_FLAG_ENABLED(grpc_backend_metric_trace)) {
+    gpr_log(GPR_INFO, "[%p] QPS utilization cleared.", this);
+  }
+}
+
+void ServerMetricRecorder::ClearNamedUtilization(absl::string_view name) {
+  std::string name_str(name);
+  internal::MutexLock lock(&mu_);
+  if (GRPC_TRACE_FLAG_ENABLED(grpc_backend_metric_trace)) {
+    gpr_log(GPR_INFO, "[%p] Named utilization cleared. name: %s", this,
+            name_str.c_str());
+  }
+  update_seq_.fetch_add(1, std::memory_order_relaxed);
+  named_utilization_.erase(name_str);
+}
+
+std::pair<grpc_core::BackendMetricData, uint64_t>
+ServerMetricRecorder::GetMetrics() const {
+  grpc_core::BackendMetricData data;
+  const double cpu = cpu_utilization_.load(std::memory_order_relaxed);
```

Comment:
Updated. I made two changes to ServerMetricRecorder:

- GetMetrics() returns optional<Data> instead of Data where nullopt indicates the data is empty and the per-call filter skips if nullopt. Making the same change to OrcaService fails tests so I instead make it shortcut to send an empty Slice.
- GetMetrics() accepts an optional last_seq from the caller, and when it matches ServerMetricRecorder skips copying out data as the caller already has data cached.
- I kept the seq atomic in ServerMetricRecorder  so that GetMetrics() can return lock-free with a cache hit. Data members are made non-atomic. So we still allow false cache misses in corner cases (but no false hits).

Reply:
Updated, I went with the shared_ptr way, I think it's worth it to reduce contention from per-call.

I had to change the ServerMetricRecorder's API to accept string_view for named fields (OrcaServer accepts std::string). I assume this is fine but lmk. Existing cases use string literals and are not affected.